---
title: '{memer}: an R interface to the MEME Suite'
author:
- ORCID: null
  name: Spencer L. Nystrom
- name: Daniel J. McKay
output: word_document
---

```{r setup}
library(dremeR)
library(magrittr)
```

# Bioconductor Abstract (300 words)

Biopolymers, such as DNA and protein, perform varying functions based on their
primary sequence. Short, repeated sequences, or "motifs" represent functional
units within biopolymers which can act as interaction surfaces, create
structure, or have enzymatic activity. Identification of shared motifs across
multiple sequences can provide evidence for shared function, thus the ability to
identify, classify, and compare motifs represents a key step in the analysis of
biological sequences. Although tools implementing these features exist within
Bioconductor, they often rely on proprietary data types with poor
interoperability with other tools. Furthermore, several of these packages will
be deprecated in the next Bioconductor release, leaving a gap in functionality
with no clear alternatives. 

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis including *de-novo* motif discovery, enrichment
testing, comparison, and matching. Here we present `memer`, a motif analysis
package which provides a seamless interface to MEME Suite tools within R.

`memer` uses standard R and Bioconductor data types for better integration with
common analysis tools like the `tidyverse` and `GRanges`. `memer` outputs also
function as inputs to other `memer` commands, allowing simple construction of motif
analysis pipelines.

In addition to wrapping MEME Suite utilities, `memer` extends the functionality
of `universalmotif` to provide base R and `tidyverse` compatible syntax for
manipulating and analyzing motif metadata. These features allow motif databases
to integrate external data to aid analysis, such as filtering a set of
transcription factor motifs based on gene expression. `memer` also provides
simple utilities for data visualization.

`memer` is designed for maximum flexibility and ease of use to allow rapid
iteration during analysis. Here we present several examples of how `memer`
allows novel analyses of ChIP and FAIRE-seq data by facilitating seamless
interoperability between MEME Suite tools and the broader R landscape.


# Abstract Draft
Biopolymers, such as DNA and protein, perform varying functions based on their
primary sequence. Short, repeated sequences, or "motifs" represent functional
units within biopolymers which can act as interaction surfaces, create
structure, or have enzymatic activity. Identification of shared motifs across
multiple sequences can provide evidence for shared function, thus the ability to
identify, classify, and compare motifs represents a key step in the analysis of
biological sequences. Although tools that implement these features have
historically existed in the Bioconductor landscape, they often rely on
proprietary data types with poor interoperability with other tools. Furthermore,
several of these packages will be deprecated in the next Bioconductor release,
leaving a gap in functionality with no clear alternatives.

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, enrichment
testing,comparison, and matching. Here we present `memer`, a motif analysis
package which provides a seamless interface to MEME Suite tools within R.

`memer` uses base R and Bioconductor data types for better integration with
common analysis tools like the `tidyverse` and other Bioconductor packages.
Unlike the commandline implementation, `memer` outputs also function as inputs
to other MEME Suite tools, allowing simple construction of motif analysis
pipelines.

In addition to wrapping MEME Suite utilities, `memer` extends the functionality
of `universalmotif` to provide base R and `tidyverse` compatible syntax for
manipulating and analyzing motif metadata. These features allow motif databases
to integrate external data to aid analysis, such as filtering a set of
transcription factor motifs based on gene expression in the tissue
of interest. `memer` also provides simple utilities for data visualization where
appropriate.

`memer` is designed for maximum flexibility and ease of use to allow users to
iterate rapidly during analysis. Here we present several examples of how `memer`
allows novel analyses of ChIP-seq and FAIRE-seq data by facilitating seamless
interoperability between MEME Suite tools and the broader R package landscape.



# Introduction

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, motif enrichment
testing, motif comparison, and motif matching, among others \cite{meme-suite}.
MEME Suite tools take as input fasta files of sequences which will be searched
for motifs, and if necessary, .meme format motif files containing motifs to use
in the search. 

Data from MEME Suite tools can be underutilized because HTML interface?
 - HTML is great human-readable reporting interface
 - `memer` allows customized reporting.

Bioconductor is deprecating PWMEnrich (~AME) and MotIV (~TomTom), `memer` can fill the gap and add additional value.

# Results 
## Design & Implementation

The key underlying design principle of `memer` was to create tools that easily
integrate into the existing suite of tools available in `Bioconductor` and the
broader `R` package landscape. In this way, `memer` tools can be used to extend
existing analysis pipelines, or to create novel analyses.

MEME Suite tools are run on the commandline and use files stored on-disk as
input while returning a series of output files which together compose the full
results of the tool. While all tools produce machine-readable data formats, few
options exist for parsing the full complement of these data into human readable
format for use outside of the MEME Suite environment.

MEME Suite are run on the commandline and take 
as input fasta files of sequences
which will be searched for motifs, and if necessary, .meme format motif files
containing motifs to use in the search (REF: table of inputs). 

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, motif enrichment
testing, motif comparison, and motif matching, among others \cite{meme-suite}.
MEME Suite tools take as input fasta files of sequences which will be searched
for motifs, and if necessary, .meme format motif files containing motifs to use
in the search. 

At its most basic form, `memer` acts as a wrapper of these functions: users can
point `memer` tools to files on disk, `memer` will construct the commandline
call, output files will be saved on disk and imported into R for further analysis.
However, `memer` is more than just a simple wrapper of the MEME Suite, it is an
analysis framework allowing seamless integration of MEME Suite tools by
providing an abstraction layer which handles file input/output (I/O) to allow
inputs and outputs in the form of in-memory R objects. 

When running a MEME Suite tool using `memer`, inputs are allowed in multiple
formats providing the user additional flexibility when performing motif analysis
(see "Core Utilities" for details). Of note, most outputs of `memer` tools can
be used as inputs to other `memer` tools, allowing easy construction of
pipelines (for example, discovering *de-novo* motifs with DREME, then finding
the genomic positions of those motifs in target regions using FIMO).
 - unlike current MEME tools (ex tomtom -> FIMO is hard to do)

By shifting the focus from managing files on disk which may change infrequently to R data
objects, this framework reinforces the requirement of writing reproducible
analysis code to generate reproducible results.

Whenever possible, `memer` uses standard R data types for better integration
with common analysis tools, like `dplyr`, `GenomicRanges`, or base R. In this
way, users don't need to learn new ways of working with data, and outputs can
be integrated with other tools with limited effort.

## Core Utilities

The core functionality of `memer` revolves around the interface to MEME Suite
tools. Each supported MEME Suite tool is assigned a `run` function
(`runDreme()`, `runMeme()`, `runAme()`, `runFimo()`, `runTomTom()`), which
accept sequence and motif inputs as needed and any valid commandline flags as
additional function parameters.

**Describe each tool briefly?**
 - This could also be relegated to a table

**table of comparison to commandline software?**
  - input type:
  - output type:

Tools requiring sequences as input can use a path to a .fasta format
file, a `Biostrings::XStringSet`, or a `Biostrings::BStringSetList` as input.
These data structures support all biological sequence types: DNA,
RNA, and amino-acid. For the remainder of the text, we will 
**use chip-seq data analysis as a framework to highlight memer features**
\strikethrough{focus our examples on DNA sequences.}

Some `memer` tools require motifs as input. In these instances, `memer` will
always accept a path to a .meme format file, or a list of motifs in
`universalmotif` format \cite{universalmotif}. The `universalmotif` package
provides simple import and motif conversion functions for all common motif data
structures and file types, ensuring cross-compatability with all commonly used
motif data formats. When `memer` functions return motifs, they are returned to
the user in `universalmotif` format. This ensures that all `memer` motif outputs
can be used as inputs to other `memer` functions that accept motif inputs.
Furthermore, `universalmotif` conversion tools allow `memer` detected motifs to
be used in many other popular Bioconductor motif packages, such as `TFBSTools`
and `motifStack`, providing the user with maximum flexibility in their analysis.

When analyzing ChIP-seq data in R, transcription factor binding locations (ChIP
Peaks) are commonly represented as `GenomicRanges` objects, which contain
genomic positions of each peak, along with an arbitrary number of metadata
columns describing features of each peak. 

A key difference between `memer` and the MEME Suite tools is that `memer`
provides tools allowing the peak-level metadata to frame subsequent analysis.
For DNA ranges (such as in ChIP-seq), the `get_sequence()` utility acts as the
entrypoint to analysis with `memer`. `get_sequence()` returns a `Biostrings`
object containing the DNA sequence of each feature in the input ranges. If
ranges are split into groups by a metadata value, `get_sequence()` will return a
list of sequences for each group. When a sequence list is used as input into
`memer` core functions, the function will run once for each group, enabling a
native "split-apply-combine" workflow without any additional user overhead
\cite{splitapplycombine, wickham}. Together, these features empower users to
use range-based metadata as the centerpoint of their analyses.

### Discriminative analysis
A powerful feature of several MEME Suite Tools is the ability to perform
discriminative analysis, where motifs are identified or tested for enrichment
in one group of sequences relative to another. In many instances, running tools
in discriminative mode on a set of control sequences outperforms simply using
shuffled input sequences as the control (back this up, change wording? maybe:
using biologically relevant control sequences can improve quality of results?).
Therefore, enabling complex discriminative analysis designs was central to the
design of `memer`.

When using list input to `memer` tools, names of list object members can be
passed to the `control` argument. This results in running the tool in
discriminative mode using those sequences as the control sequences for all other
list members. By passing multiple names to `control`, the sequences are combined
together to create a merged background set. In this way, performing data-driven
discriminative analysis is just as simple as one using shuffled sequences as
background.

```{r, eval=F}
############
# Replace this for a real set of peaks 
sequences_list <- list("Static" = "seq",
                       "Increasing" = "seq",
                       "Decreasing" = "seq")
############

# run Dreme on Static, Increasing, Decreasing sites using shuffled input as background
runDreme(sequences_list, "shuffle")

# run Dreme on Increasing, Decreasing sites using Static sites as background
runDreme(sequences_list, "Static")

# run Dreme on Decreasing sites using a combined set of Static + Increasing sites as background
runDreme(sequences_list, c("Static", "Increasing"))
```

## Metadata manipulation
`memer` is built on top of the `universalmotif` framework, which provides
methods for importing motifs from various file formats into a unified R data
representation complete with metadata information \cite{universalmotif}.
However, a major limitation of the `universalmotif` framework is that 
manipulating or filtering multiple `universalmotif` entries requires potentially
unfamiliar overhead for less experienced users, acting as a barrier to
performing potentially useful operations.

The `as_universalmotif_dataframe()` function from `memer` converts
`universalmotif` objects into a more familiar `data.frame` format which can be
manipulated using `tidyverse` or base R tools. These data.frames also contain a
special `motif` column which stores the `universalmotif` object of each entry.
Users can update columns in the data.frame which will be reflected in the
corresponding metadata entries of the `motif` by running `update_motifs()` after
manipulating the columns. These data.frames can be converted back into
`universalmotif` format using `as_universalmotif`, which also updates the
metadata using column values. The results from running the core MEME Suite
functions are stored in this format whenever appropriate, creating a consistent
framework for users to explore and manipulate motifs and their associated data values.

**This section is maybe too listy and could be fleshed out more / integrated above better**
Examples of common manipulations to motif databases which are aided by this
format include bulk renaming of entries, identification and removal of redundant
entries, and exploratory analysis of motif databases.

```{r}
# Example of renaming motifs w/ universalmotif syntax
library(universalmotif)
m1 <- create_motif()
m2 <- create_motif()
my_motifs <- list(m1,m2)

my_names <- c("motif one", "motif two")

for (i in 1:length(my_motifs)) {
  my_motifs[[i]]@name <- my_names[i]
}

# Example of renaming motifs w/ memer syntax
my_motifs <- list(m1,m2)
my_motifs %<>% 
  as_universalmotif_dataframe() %>% 
  dplyr::mutate(name = my_names) %>% 
  as_universalmotif()
```

This data format also allows users to integrate external data sources into their
analyses, for example, to filter a motif database to include only motifs for
genes which are expressed in their tissue of interest. Using motif databases
filtered in this way has been demonstrated in multiple contexts to provide
higher quality results when searching for novel regulators based on motif
content \cite{hector franco, eric olson, john stamm?, shep & greenleaf, klu worm
paper}.

## Data Visualization (section unfinished)

The MEME Suite provides a limited set of data visualizations which have limited
customizability. `memer` leverages the advantages of the R graphics environment
to provide novel data visualizations which are infinitely customizable by the user.

  - `view_tomtom_hits()` allows examination of tomtom results, allowing the user
  to evaluate the quality of the assigned match.

```{r}
# Example comparison of view_tomtom_hits results
plot_topHits <- example_tomtom %>% 
  dplyr::mutate(name = "Example Motif (Has Matches)") %>% 
  update_motifs() %>% 
  view_tomtom_hits(2)

plot_noMatch <- create_motif() %>% 
  as_universalmotif_dataframe() %>% 
  dplyr::mutate(name = "Example Motif (No Matches)") %>% 
  update_motifs() %>% 
  {
    x <- .
    x$tomtom <- list(NULL)
    return(x)
  } %>% 
  view_tomtom_hits()

cowplot::plot_grid(plotlist = list(plot_topHits[[1]], plot_noMatch[[1]]), ncol = 2, labels = "AUTO")
```
 
  - `ame_plot_heatmap()` allows comparison of AME results between multiple
  groups. Heirarchical ordering of hits shows commonalities and differences in enrichment between groups.
```{r}
# AME Heatmap example plots
example_ame$Decreasing %>% 
  ame_plot_heatmap()

example_ame %>% 
  dplyr::bind_rows(.id = "id") %>% 
  ame_plot_heatmap(group = id)
```

# Conclusion
 - `memer` is cool
 - use `memer`
 - it's on github (provide link)
 - manual is here:

# Notes

**Example of commandline vs `memer`?**
For example, to run DREME on the commandline on "input.fasta", while setting the random seed to 100:
```{bash, eval=F}
# Shell command setting random seed to 100
dreme -s 100 input.fasta
```

The `memer` implementation of the above command would be written as follows:
```{r, eval=F}
# memer command setting random seed to 100
# user explicitly states they want to shuffle input sequences as control
runDreme("input.fasta", control = "shuffle", s = 100)
```

### Motif Comparison (TomTom)
 - takes runDreme/meme output as input -> pipes
 - enables using best matched motif as input to other tools (unlike TomTom on commandline)

### Motif Matching (FIMO)
 - returns locations as GRanges
 - integrates well with `plyranges` to facilitate range-based join operations to integrate peak-level metadata with motif information, vice-versa
 

\begin{figure}
\includegraphics[width=\textwidth]{figures/memer_model.png}
\caption{A diagram of memer tools & workflows?
flowchart of each step, getsequence-> run-> import -> visualize -> goto run

Diagram how meme-suite tools work vs memer?
}
\label{fig:model} 
\end{figure}


  
# Future developments:
 - Additional meme suite tools
 - better support for PSP generation
 - Better support for amino-acid sequence usage (please contact us with feature requests)