---
title: "{memer}: an R interface to the MEME Suite"
author: 
 - name: Spencer L. Nystrom
   ORCID: 
 - name: Daniel J. McKay
output: github_document
---

```{r setup}
library(dremeR)
```

# Introduction

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, motif enrichment
testing, motif comparison, and motif matching, among others \cite{meme-suite}.
MEME Suite tools take as input fasta files of sequences which will be searched
for motifs, and if necessary, .meme format motif files containing motifs to use
in the search. 

Data from MEME Suite tools can be underutilized because HTML interface?
 - HTML is great human-readable reporting interface
 - `memer` allows customized reporting.

Bioconductor is deprecating PWMEnrich (~AME) and MotIV (~TomTom), `memer` can fill the gap and add additional value.

# Results 
## Design & Implementation

The key underlying design principle of `memer` was to create tools that easily
integrate into the existing suite of tools available in `Bioconductor` and the
broader `R` package landscape. In this way, `memer` tools can be easily integrated
into existing analysis pipelines, or used to create novel analyses. 

MEME Suite tools are run on the commandline and use files stored on-disk as
input while returning a series of output files which together compose the full
results of the tool. While all tools produce machine-readable data formats, few
options exist for parsing the full complement of these data into human readable
format for use outside of the MEME Suite environment.

MEME Suite are run on the commandline and take 
as input fasta files of sequences
which will be searched for motifs, and if necessary, .meme format motif files
containing motifs to use in the search (REF: table of inputs). 

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, motif enrichment
testing, motif comparison, and motif matching, among others \cite{meme-suite}.
MEME Suite tools take as input fasta files of sequences which will be searched
for motifs, and if necessary, .meme format motif files containing motifs to use
in the search. 

At its most basic form, `memer` acts as a wrapper of these functions: users can
point `memer` tools to files on disk, `memer` will construct the commandline
call, output files will be saved on disk and imported into R for further analysis.
However, `memer` is more than just a simple wrapper of the MEME Suite, it is an
analysis framework allowing seamless integration of MEME Suite tools by
providing an abstraction layer which handles file input/output (I/O) to allow
inputs and outputs in the form of in-memory R objects. 

When running a MEME Suite tool using `memer`, inputs are allowed in multiple
formats providing the user additional flexibility when performing motif analysis
(see "Core Utilities" for details). Of note, most outputs of `memer` tools can
be used as inputs to other `memer` tools, allowing easy construction of
"pipelines" (for example, discovering *de-novo* motifs with DREME, then finding
the genomic positions of those motifs in target regions using FIMO).

By shifting the focus from managing files on disk which may change infrequently to R data
objects, this framework reinforces the requirement of writing reproducible
analysis code to generate reproducible results.

Whenever possible, `memer` uses standard R data types for better integration
with common analysis tools, like `dplyr`, `GenomicRanges`, or base R. In this
way, users don't need to learn new ways of working with data, and outputs can
be integrated with other tools with limited effort.


### Detecting a local MEME Suite Install (remove this section?)
 
`memer` requires the user to install the MEME Suite to the analysis environment
in which `memer` will be used. The `check_meme_install()` function allows users
to determine whether their R session is able to detect a local install of the
MEME Suite.
  - For functions that use MEME commandline tools, they need the location of 'meme/bin/' on disk, we provide utilities for making this detection simple to reduce friction of using the tool.
    - `check_meme_install` does user-friendly check that meme can be detected from R environment
    - Environment variable `MEME_PATH` (set & forget), `options(meme_path)` (configure for each script), `meme_path` argument, for each function call

## Core Utilities

The core functionality of `memer` revolves around the interface to MEME Suite
tools. Each supported MEME Suite tool is assigned a `run` function
(`runDreme()`, `runMeme()`, `runAme()`, `runFimo()`, `runTomTom()`), which
accept sequence and motif inputs as needed and any valid commandline flags as
additional function parameters.

Tools requiring sequences as input can use a path to a .fasta format
file, a `Biostrings::XStringSet`, or a `Biostrings::BStringSetList` as input.
These data structures support all biological sequence types: DNA,
RNA, and amino-acid. For the remainder of the text, we will 
**use chip-seq data analysis as a framework to highlight memer features**
\strikethrough{focus our examples on DNA sequences.}

For DNA and RNA sequences, the `get_sequence()` utility acts
as the entrypoint to analysis with `memer`. `get_sequence()` returns
`Biostrings::XStringSet` or `Biostrings::BStringSetList` types from range-based
data types such as `GenomicRanges` or `GenomicRangesList` to serve as
sequence input to the `memer` core functions.

When analyzing ChIP-seq data in R, transcription factor binding locations (ChIP
Peaks) are commonly represented as `GenomicRanges` objects, which contain
genomic positions of each peak, along with an arbitrary number of metadata
columns describing features of each peak. 

Some `memer` tools require motifs as input. In these instances, `memer` will
always accept a path to a .meme format file, or a list of motifs in
`universalmotif` format \cite{universalmotif}. The `universalmotif` package
provides simple import and motif conversion functions for all common motif data
structures and file types, ensuring cross-compatability with all commonly used
motif data formats. When `memer` functions return motifs, they are returned to
the user in `universalmotif` format. This ensures that all `memer` motif outputs
can be used as inputs to other `memer` functions that accept motif inputs.
Furthermore, `universalmotif` conversion tools allow `memer` detected motifs to
be used in many other popular Bioconductor motif packages, such as `TFBSTools`
and `motifStack`, providing the user with maximum flexibility in their analysis.

**Example of commandline vs `memer`?**
For example, to run DREME on the commandline on "input.fasta", while setting the random seed to 100:
```{bash, eval=F}
# Shell command setting random seed to 100
dreme -s 100 input.fasta
```

The `memer` implementation of the above command would be written as follows:
```{r, eval=F}
# memer command setting random seed to 100
# user explicitly states they want to shuffle input sequences as control
runDreme("input.fasta", control = "shuffle", s = 100)
```

**Somewhere say "memer can be used as a simple wrapper, but here we highlight its utility in chip-seq data analysis".**

### Analysis using multiple groups (split-apply-combine approach)
Users can adopt a "split-apply-combine" strategy for using `memer`. `GRanges`
object acts as the centerpoint, can be split by data, fed to `get_sequence()`,
then the list will run the tools for each group automatically.

### Discriminative analysis
A key feature of several MEME Suite Tools is the ability to perform
discriminative analysis, where motifs are identified or tested for enrichement
in one group of sequences relative to another. In many instances, running tools
in discriminative mode on a set of control sequences outperforms simply using
shuffled input sequences as the control (back this up, change wording?). Therefore, 
enabling complex discriminative analysis designs was central to the design of `memer`.

When using list input to `memer` tools, names of list object members can be
passed to the `control` argument. This will result in running the tool in
discriminative mode using those sequences as the control sequences for all other
list members. By passing multiple names to `control`, the sequences will be
combined together to create a background set.

```{r}
sequences_list <- list("Static" = "seq",
                       "Increasing" = "seq",
                       "Decreasing" = "seq")

# run Dreme on Static, Increasing, Decreasing sites using shuffled input as background
runDreme(sequences_list, "shuffle")

# run Dreme on Increasing, Decreasing sites using Static sites as background
runDreme(sequences_list, "Static")

# run Dreme on Decreasing sites using a combined set of Static + Increasing sites as background
runDreme(sequences_list, c("Static", "Increasing"))
```

### Motif Comparison (TomTom)
 - takes runDreme/meme output as input -> pipes

### Motif Matching (FIMO)
 - returns locations as GRanges
 - integrates well with `plyranges` to facilitate range-based join operations to integrate peak-level metadata with motif information, vice-versa
 

\begin{figure}
\includegraphics[width=\textwidth]{figures/memer_model.png}
\caption{A diagram of memer tools & workflows?
flowchart of each step, getsequence-> run-> import -> visualize -> goto run

Diagram how meme-suite tools work vs memer?
}
\label{fig:model} 
\end{figure}


  
# Future developments:
 - Additional meme suite tools
 - better support for PSP generation
 - Better support for amino-acid sequence usage (please contact us with feature requests)