---
title: '{memes}: an R interface to the MEME Suite'
author:
- ORCID: null
  name: Spencer L. Nystrom
- name: Daniel J. McKay
output: word_document
---

```{r setup}
library(memes)
library(magrittr)
```

# Bioconductor Abstract (300 words)

Biopolymers, such as DNA and protein, perform varying functions based on their
primary sequence. Short, repeated sequences, or "motifs" represent functional
units within biopolymers which can act as interaction surfaces, create
structure, or have enzymatic activity. Identification of shared motifs across
multiple sequences can provide evidence for shared function, thus the ability to
identify, classify, and compare motifs represents a key step in the analysis of
biological sequences. Although tools implementing these features exist within
Bioconductor, they often rely on proprietary data types with poor
interoperability with other tools. Furthermore, several of these packages will
be deprecated in the next Bioconductor release, leaving a gap in functionality
with no clear alternatives. 

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis including *de-novo* motif discovery, enrichment
testing, comparison, and matching. Here we present `memes`, a motif analysis
package which provides a seamless interface to MEME Suite tools within R.

`memes` uses standard R and Bioconductor data types for better integration with
common analysis tools like the `tidyverse` and `GRanges`. `memes` outputs also
function as inputs to other `memes` commands, allowing simple construction of motif
analysis pipelines.

In addition to wrapping MEME Suite utilities, `memes` extends the functionality
of `universalmotif` to provide base R and `tidyverse` compatible syntax for
manipulating and analyzing motif metadata. These features allow motif databases
to integrate external data to aid analysis, such as filtering a set of
transcription factor motifs based on gene expression. `memes` also provides
simple utilities for data visualization.

`memes` is designed for maximum flexibility and ease of use to allow rapid
iteration during analysis. Here we present several examples of how `memes`
allows novel analyses of ChIP and FAIRE-seq data by facilitating seamless
interoperability between MEME Suite tools and the broader R landscape.


# Abstract Draft
Biopolymers, such as DNA and protein, perform varying functions based on their
primary sequence. Short, repeated sequences, or "motifs" represent functional
units within biopolymers which can act as interaction surfaces, create
structure, or have enzymatic activity. Identification of shared motifs across
multiple sequences can provide evidence for shared function, thus the ability to
identify, classify, and compare motifs represents a key step in the analysis of
biological sequences. Although tools that implement these features have
historically existed in the Bioconductor landscape, they often rely on
proprietary data types with poor interoperability with other tools. Furthermore,
several of these packages will be deprecated in the next Bioconductor release,
leaving a gap in functionality with no clear alternatives.

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, enrichment
testing, comparison, and matching. Here we present `memes`, a motif analysis
package which provides a seamless interface to MEME Suite tools within R.

`memes` uses base R and Bioconductor data types for better integration with
common analysis tools like the `tidyverse` and other Bioconductor packages.
Unlike the commandline implementation, `memes` outputs also function as inputs
to other MEME Suite tools, allowing simple construction of motif analysis
pipelines.

In addition to wrapping MEME Suite utilities, `memes` extends the functionality
of `universalmotif` to provide base R and `tidyverse` compatible syntax for
manipulating and analyzing motif metadata. These features allow motif databases
to integrate external data to aid analysis, such as filtering a set of
transcription factor motifs based on gene expression in the tissue
of interest. `memes` also provides simple utilities for data visualization where
appropriate.

`memes` is designed for maximum flexibility and ease of use to allow users to
iterate rapidly during analysis. Here we present several examples of how `memes`
allows novel analyses of ChIP-seq and FAIRE-seq data by facilitating seamless
interoperability between MEME Suite tools and the broader R package landscape.



# Introduction

The MEME Suite is a series of commandline tools which provide a variety of
features for motif analysis such as *de-novo* motif discovery, motif enrichment
testing, motif comparison, and motif matching, among others \cite{meme-suite}.
MEME Suite tools take as input fasta files of sequences which will be searched
for motifs, and if necessary, .meme format motif files containing motifs to use
in the search. 

Data from MEME Suite tools can be underutilized because HTML interface?
 - HTML is great human-readable reporting interface
 - `memes` allows customized reporting.

Bioconductor is deprecating PWMEnrich (~AME) and MotIV (~TomTom), `memes` can fill the gap and add additional value.




Biopolymers, such as DNA and protein, perform varying functions based on their
primary sequence. Short, repeated sequences, or "motifs" represent functional
units within biopolymers which can act as interaction surfaces, create
structure, or have enzymatic activity. Identification of shared motifs across
multiple sequences can provide evidence for shared function, thus the ability to
identify, classify, and compare motifs represents a key step in the analysis of
biological sequences. Although tools that implement these features have
historically existed in the Bioconductor landscape, they often rely on
proprietary data types with poor interoperability with other tools. Furthermore,
several of these packages will be deprecated in the next Bioconductor release,
leaving a gap in functionality with no clear alternatives.

The MEME Suite is a series of commandline and web-based tools which provide a variety
of features for motif analysis such as *de-novo* motif discovery, enrichment
testing, comparison, and matching \cite{meme-suite}. The MEME Suite is a widely
accepted set of tools with over 22,000 citations to date, and over 30,000 unique
users of the webserver implementation annually \cite{meme-usage-report-page,
google-scholar-meme}. 

The MEME Suite family of tools provides a
comprehensive set of utilities to interrogate motif content across a broad range
of biological contexts. However, several factors limit the full potential of
these tools for use in data analysis. 
  1) must do significant preprocessing to get data in correct input format
    - written to analyze 1 group of sequences, not amenable to multigroup designs
    - no common patterns for generating inputs from ChIP-seq whose data are primarily range based
  2) Easily parseable output data often doesn't contain all information
    - ex tomtom alternative hits in .tsv output only present in IUPAC format, not matrix
      - matrix is present in XML file, but no tools exist to parse this
  3) limited data visualization capabilities
  4) difficult to integrate data with additional sources

Here we present `memes`, a motif analysis package which provides a seamless
interface to MEME Suite tools within R. `memes` uses base R and Bioconductor
data types for data input and output, facilitating better integration with
common analysis tools like the `tidyverse` and other Bioconductor packages.
Unlike the commandline implementation, `memes` outputs also function as inputs
to other MEME Suite tools, allowing simple construction of motif analysis
pipelines without additional data processing steps. Additionally, R/Bioconductor
data structures provide a full-featured representaiton of MEME Suite output
data, providing users quick access to all relevant data structures with simple
syntax. 

In addition to wrapping MEME Suite utilities, `memes` extends the functionality
of `universalmotif` to provide base R and `tidyverse` compatible syntax for
manipulating and analyzing motif metadata. These features allow motif databases
to integrate external data to aid analysis, such as filtering a set of
transcription factor motifs based on gene expression in the tissue of interest.

`memes` also provides simple utilities for data visualization where appropriate.
Providing R representations of MEME Suite data objects enables generation of
novel data visualizations using the well established tools available in the R
landscape. Therefore, `memes` empowers users to deeply analyze and visualize
motif analysis data.

`memes` is designed for maximum flexibility and ease of use to allow users to
iterate rapidly during analysis. Here we present several examples of how `memes`
allows novel analyses of ChIP-seq and FAIRE-seq data by facilitating seamless
interoperability between MEME Suite tools and the broader R package landscape.

# Results 
## Design & Implementation

### Core Utilities

The core functionality of `memes` revolves around the interface to MEME Suite
tools. Each supported MEME Suite tool is assigned a `run` function
(`runDreme()`, `runMeme()`, `runAme()`, `runFimo()`, `runTomTom()`), which
accept sequence and motif inputs as needed and any valid commandline flags as
additional function parameters.

**Describe each tool briefly?**
 - This could also be relegated to a table

**table of comparison to commandline software?**
  - input type:
  - output type:
  
### Data Structures

MEME Suite tools are run on the commandline and use files stored on-disk as
input while returning a series of output files containing varying data types.
`memes` reduces each MEME Suite tool to an R function which internally handles
file I/O such that all MEME Suite calling functions accept R objects as input,
and return R objects as output. Sequence inputs are stored in `BioStrings`
format, an R/Bioconductor package for storing biopolymer sequence data
\cite{BioStrings}. Motif inputs are passed as `universalmotif` objects, another
R/Bioconductor package for representing motif matrices and their associated
metadata \cite{universalmotif}. `memes` will also always accept paths to files
on disk, such as fasta files for sequence inputs, and meme format files for
motif inputs, reducing the need to read large files into memory to run MEME
Suite tools.

Although MEME Suite tools accept consistent data type inputs, they do not return
consistent data structures across tools. For example, the tool TomTom, which
compares query motifs to a database of known motifs, returns tab-separated
results which do not contain the frequency matrix representations of the matched
motifs. Instead, users must write custom code to parse these matrices back out
from the input databases, creating additional friction for creating novel data
visualizations and exploratory analyses. Often, output data are also returned in
XML format, which although they hold all relevant information, are difficult to
parse. To our knowledge, no publicly available tools exist to parse MEME Suite
XML formatted data, therefore users must write custom code to extract these
data. `memes` exports data import functions for each supported MEME Suite tool,
which import these data as R data.frames. These functions also work internally to
create the seamless file I/O when calling MEME Suite tools, but the import
functions can be called directly by users to import previously generated data
for use in R.

 - **NOTE: Biopython has a MEME/MAST parser but it works on tsv https://biopython.org/docs/1.74/api/Bio.motifs.meme.html**
 - **Actually, they merged XML import in March: https://github.com/biopython/biopython/pull/2354**
 - **Need to revise above statement, but note that most tools don't have support**

Many of the MEME Suite tools, such as MEME, DREME, or TomTom return data which
describe properties of motifs. For example, MEME and DREME return de-novo
discovered motifs from query sequences along with statistical information about
their enrichment. The `universalmotif` data structure is a powerful
R/Bioconductor representation for motif matrices and their associated metadata \cite{universalmotif}. 
However, `universalmotif` objects are not ideal for performing bulk manipulation
or queries of metadata, nor do they support complex data structures such as
nested elements. Many MEME Suite analysis pipelines involve identifying motifs
which are further analyzed by multiple tools, creating additional metadata which
cannot be easily tracked in bulk using `universalmotif` format. Thus
an ideal representation for these data is one which can hold an unlimited amount
of metadata, contain complex heirarchies, and be easily manipulated using
standard analysis tools. `memes` therefore extends the `universalmotif` format
by providing data.frames holding motif metadata which retain links to their
`universalmotif` representation. These universalmotif
data.frames combine the advantages of the `universalmotif` framework for motif
manipulation with the extensibility and flexibility of R data.frames. These
`universalmotif` data.frames form the basis for a majority of `memes` data
outputs. These structures are also valid input types to `memes` functions, and
when used as such, output data are appended as new columns to the input data,
ensuring data provenance.

**Figure?: showing how the data.frame connects to universalmotif**

`universalmotif` data.frames are not limited to storing motifs from MEME Suite
data, and can additionally be used to examine or manipulate the properties of
any motif list or database. Using `universalmotif` functions to import motif data,
the `memes` function `as_universalmotif_dataframe()` converts any `universalmotif` object
into a `universalmotif` data.frame which can be manipulated using base R or
`tidyverse` functions, just like regular R data.frames. Manipulation of linked
metadata columns will result in updating the linked `universalmotif` object when
calling `update_motifs()` to retain the data.frame structure, or when calling
`as_universalmotif()` to convert the data.frame back into `universalmotif` format. In
this way, `memes` enables large-scale motif database manipulation & analysis
within the preexisting R framework.

### Support for genomic range-based data

# RETURN
Motif analysis is important for assays like ChIP-seq which store data as genomic
coordinates, not sequence. This is in conflict with MEME Suite which is designed
to study sequences. Therefore, users must write custom code to extract DNA
sequence of their ranges of interest. Although tools like `bedtools getfasta`
can simplify the process of extracting DNA sequence from genomic coordinates,
some MEME tools require fasta headers to be specially formatted to extract full
functionality, requiring users to write additional custom code to properly
format input data. 

`memes` provides the `get_sequence` function which automates the extraction of
DNA sequence from genomic coordinates while simultaneously producing MEME Suite
formatted fasta headers. `get_sequence` accepts genomic-range based inputs in
`GenomicRanges` format, the *de-facto* standard for genomic coordinate
representation in R \cite{genomicranges}. Other common genomic coordinate
representations, such as bed format, are easily imported as `GenomicRanges`
objects into R using preexisting import functions, meaning users must not write
any custom import functions to work with range-based data using `memes`. 
Sequences are returned in `Biostrings` format and can therefore be used as
inputs to all `memes` commands or as inputs to other R/Bioconductor functions
for sequence analysis. 

During analysis of multiomic data, users often uncover different functional
categories of genomic regions through integration of multiple datasets. In these
instances, motif analysis acts as a powerful tool to examine whether differences
in motif content is associated with membership of distinct categories. In some
cases, performing differential motif analysis (using one category as the "test"
set and another as the "control") can uncover biologically relevant motifs which
are enriched in one category relative to another. Although the MEME Suite allows
differential enrichment testing, it does not inheritly provide a mechanism for
analyzing groups of sequences in parallel, or performing motif analysis with any
understanding of data categories. The `memes` framework enables "data-aware"
motif analysis workflows by allowing named lists of `Biostrings` objects as
input to each function. If using `GenomicRanges`, users can split peak data on a
metadata column using the base R `split` function, and use this result as input
to `get_sequence`, which will produce a list of `BioStrings` objects where each
entry is named after each data category from the split column. When a list is
used as input to a `memes` function, it runs the corresponding MEME Suite tool
for each object in the list. Users can also pass the name(s) of a category to
the `control` argument to automatically enable differential analysis of all
other list members against the control category sequences.  Therefore, `memes`
enables complex data-aware differential motif analysis workflows using simple
syntax to extend the capabilities of the MEME Suite.

### Table of supported MEME Suite Tools
 - comparison to other tools?
 - Maybe only do this if asked

## Data visualization
The MEME Suite provides a small set of data visualizations which have limited
customizability. `memes` leverages the advantages of the R graphics environment
to provide novel data visualizations which are infinitely customizable by the user. 

 - `view_motifs()` plots motif as stack (if I decide to add this feature... need to think about this)
 - `ame_plot_heatmap` plot clustered heatmap of AME results 
   - `ame_compare_heatmap_methods` compare nonlinearity of data visualization in heatmap form
 - `view_tomtom_hits` visualize and compare TomTom matches to assess match quality
   - `force_best_match` can be used to update "best" match demonstrating how data visualization can be useful for data curation

The AME tool searches for enrichment of motifs within groups of sequences
relative to a control set. The meaningful result from this tool is the
statistical parameter (for example, a p-value) associated with the significance
of motif enrichment. However, AME does not provide a mechanism for visualizing
these results. Additionally, AME is often used 
## RETURN


## Containerized analysis maximizes availability and facilitates reproducibility
`memes` relies on a locally installed version of the MEME Suite which is
accessed by the users R process. Although R is available on Windows, Mac, and
Linux operating systems, the MEME Suite is incompatible with Windows, limiting
its adoption by Windows users. Additional barriers also exist to installing a
local copy of the MEME Suite on compatible systems. The MEME Suite relies on
several system-level dependencies whose installation can be difficult for novice
users. Finally, some tools in the MEME Suite use python to generate shuffled
control sequences for analysis, which presents a reproducibility issue as the
random number generation algorithm changed from python2.7 to python3. The MEME
Suite will build and install on both python2.7 and python3 systems, therefore
without careful consideration of this behavior, the same code run on two systems
may not produce identical results, even if using the same major version of the
MEME Suite. In order to democratize access to the MEME Suite on unsupported
operating systems, and to facilitate reproducibile motif analysis using `memes`, 
we have also developed docker and singularity (**NOTE:** singularity is broken
for `rocker/rstudio`, so maybe omit this until fixed) containers containing a
preinstalled version of the MEME Suite along with an R/Bioconductor analysis
environment including the most recent version of `memes` and its dependencies.

# Results

Here we briefly highlight each of `memes` current features for analyzing
ChIP-seq data. Additional detailed walkthroughs for each supported MEME Suite
tool, and a complete example using `memes` to analyze ChIP-seq data can be found
in the `memes` vignette and the package website
([snystrom.github.io/memes](snystrom.github.io/memes)). In the following
example, we reanalyze recent work examining the causal relationship between the
binding of the transcription factor E93 to changes in chromatin accessibility
during *Drosophila* wing development \cite{Nystrom, Niederhuber 2020}. Here, we
utilize ChIP-seq peak calls for E93 which have been annotated according to the
change in chromatin accessibility observed before & after E93 binding. These
data are an ideomatic example of range-based genomic data (E93 ChIP peaks)
containing additional groupings (the chromatin accessibility response following
DNA binding) whose membership may be influenced by differential motif content.
We show how `memes` syntax allows complex analysis designs, how `memes`
utilities enable deep interrogation of results, and how `memes` flexible data
structures empower users to integrate the `memes` workflow with tools offered by
other R/Bioconductor packages.

### RETURN HERE
### NOTE TO SELF:
 - This is too much of a tutorial & vignette.
 - move this text to the vignette since there isn't any.
 - Instead, highlight some of the interface useage like so:

To run dreme, where `peaks` are chip peaks below with a metadata column `e93_respnose`:
```{r}
seq <- peaks %>% split(mcols(.)$response) %>% get_sequence
runDreme(seq, control = "static")
```

----------------
```{r, include=FALSE}
drake::loadd(chip_results)
```

We begin with a `GenomicRanges` object of E93 ChIP-peak summits +/- 100 bp with
a metadata column indicating the change in chromatin accessibility associated
with E93 binding at that region. 
```{r}
chip_results %>% head
```

### Denovo motif discovery using DREME

One approach to identify *de-novo* motifs within ChIP peaks is to use DREME,
which performs discriminative motif detection in an input set of sequences
relative to a control set \cite{dreme}. To search for *de-novo* motifs within
E93 bindings sites using DREME, we first use the `get_sequence()` function which
returns the DNA sequence within genomic coordinates.
```{r}
genome <- BSgenome.Dmelanogaster.UCSC.dm3::BSgenome.Dmelanogaster.UCSC.dm3

all_seq <- chip_results %>% 
  get_sequence(genome)
```

The output is stored as a `DNAStringSet`, shown below.
```{r}
head(all_seq, 3)
```

The `runDreme` command is used to begin a motif search using DREME. Although it
is possible to use any number of input sequences, runtimes quickly become
infeasible as sequences exceed 100000 total bp. This can be overcome in a
variety of ways: users can rank peaks by ChIP-seq signal, or some other
quantitative value, and use a subset selected by rank, or in the absence of any
quantitative measure, users can randomly sample their sequences to use as input.
For simplicity in this example, we subset our data to 1000 randomly selected
peaks. Setting `control = "shuffle"` uses DREME's built-in dinucleotide
shuffling algorithm to produce a control set of seqeunces.

```{r}
set.seed(123) # setting a random seed ensures sampling produces a reproducible subsample
sequence_subset <- sample(all_seq, 1000)
all_dreme_res <- runDreme(sequence_subset, control = "shuffle", nmotifs = 3)
```

```{r}
all_dreme_res
```



### Preparing a motif database for searching

Transcription factors often bind with sequence specificity to regulate activity
of nearby genomic elements. Therefore, comparison of *de-novo* motifs with known
DNA binding motifs for transcription factors can be a key step in identifying
regulators of genomic elements.

The TomTom tool is used to compare unknown motifs against a list of known motifs
to identify matches. By passing the results from DREME into TomTom and searching
within a database of *Drosophila* transcription factor motifs, we can identify
candidate transcription factors which may also bind at E93 ChIP peaks. A major
limitation of this approach is that transcription factors containing similar
families of DNA binding domain often possess highly similar motifs, making it
difficult to identify the "true" binding factor of a discovered motif.
Additionally, when searching for matches against a motif database, TomTom must
account for multiple testing, therefore using a larger than necessary motif
database can produce a large multiple testing penalty, limiting sensititity of
detection. One way to overcome these limitations is to limit the transcription
factor motif database to include only motifs for transcription factors expressed
in the sample of interest. Accounting for transcription factor expression during
motif analysis has been demonstrated to  increase the probability of identifying
biologically relevant transcription factor candidates \cite{hector franco's STAR paper, others?}.

The `memes` universalmotif data.frame structure can be used to integrate
expression data with a motif database. First, we import a pre-filtered list of
genes expressed in a timecourse of **Drosophila** wing development.

```{r}
rna_counts <- read.csv("data/wing_expressed_genes.csv", stringsAsFactors = FALSE)
```
```{r, echo = FALSE}
head(rna_counts)
```

Next we import a database of known *Drosophila* transcription factor motifs
generated from the Fly Factor Survey in `universalmotif` format, and convert to a data.frame format using `as_universalmotif_dataframe()` \cite{fly factor}.
```{r}
flyFactor <- universalmotif::read_meme("data/flyFactorSurveyCleaned.meme") %>% 
  as_universalmotif_dataframe()
```
```{r, echo = FALSE}
head(flyFactor, 5)
```


```{r}
tomtom_results <- runTomTom(all_dreme_res, database = "data/flyFactorSurveyCleaned.meme", dist = "ed")
```
Here we visualize the top 3 matches from TomTom for each *de-novo* motif using
`view_tomtom_hits()`. Results are plotted in descending order according to match
score. The results of this function can be integrated into additional R graphics
libraries, such as `patchwork` or `cowplot` to generate publication-quality
figures \cite{patchwork, cowplot}.
```{r, fig.height=5, fig.width=12}
tomtom_results %>% 
  view_tomtom_hits(3) %>% 
  patchwork::wrap_plots(., nrow = 1, tag_level = "new") +
    patchwork::plot_annotation(tag_levels = "A")
```
## Replace above chunks with sections below 
(rstudio crashed without saving)
--------------
Next we import a database of known *Drosophila* transcription factor motifs
generated from the Fly Factor Survey in `universalmotif` format, and convert to a data.frame format using `as_universalmotif_dataframe()` \cite{fly factor}.
```{r}
fly_factor <- universalmotif::read_meme("data/flyFactorSurveyCleaned.meme") %>% 
				  as_universalmotif_dataframe()
	```
	```{r, echo = FALSE}
	head(fly_factor, 5)
	```

The motif database can be subsequently filtered to include only those entries
which match an expressed gene. After filtering, the database is converted back
to `universalmotif` format using `as_universalmotif`.
```{r}
# in this database, altname stores the gene symbol
fly_factor_expressed <- fly_factor %>% 
				  dplyr::filter(altname %in% expressed_genes$symbol) %>% 
					  as_universalmotif()
		```

Passing the results object from *de-novo* discovery with `runDreme` and the
filtered motif database to `runTomTom`
```{r}
tomtom_results <- runTomTom(all_dreme_res, database = fly_factor_expressed, dist = "ed")
tomtom_results <- runTomTom(all_dreme_res, database = fly_factor, dist = "ed")
```

----------------




 - prepare database
  - filter FPKM counts, note this needs to be chosen empirically and is not always a good idea,
  - we do this for demonstration purposes only
 - Denovo discovery in sensitivity categories (static as background)
  - TOMTOM matching, view_match_motifs(3)
    - try patchwork to combine them into figure
 - AME Across binding groups
  - reduce redundant hits
  - plot heatmap
 - FIMO within sensitivity
  - rederive PWM for E93 within sites
   - note differences in specific bases

  
# Availability and Future developments:
 - packge source code: `github.com/snystrom/memes`
 - documentation website: `snystrom.github.io/memes`
 - docker container: `https://hub.docker.com/r/snystrom/memes_docker`
 - singularity container: `???`

### Future Features:
 - Additional meme suite tools
 - more data visualizations
 - better support for PSP generation
 - Better support for amino-acid sequence usage (please contact us with feature requests)
 
# Acknowledgements
 - Sean Davis automated docker builds
 - Megan Justice, feedback, feature requests
